akka-persistence/src/main/scala/akka/persistence/Persistent.scala

```diff
@@ -36,20 +36,24 @@ object AtomicWrite {
 
 final case class AtomicWrite(payload: immutable.Seq[PersistentRepr]) extends PersistentEnvelope with Message {
   require(payload.nonEmpty, "payload of AtomicWrite must not be empty!")
+  private var optHighestSequenceNr: Option[Long] = None
 
   // only check that all persistenceIds are equal when there's more than one in the Seq
   if (payload match {
     case l: List[PersistentRepr]   ⇒ l.tail.nonEmpty // avoids calling .size
     case v: Vector[PersistentRepr] ⇒ v.size > 1
     case _                         ⇒ true // some other collection type, let's just check
   }) require(
-    payload.forall(_.persistenceId == payload.head.persistenceId),
+    payload.forall { pr ⇒
+      optHighestSequenceNr = Some(pr.sequenceNr)
```

Comment:
as @ktoso suggested, the `Option` must go away, because this is allocating new instances, which is probably worse than the original double traversal